### PR TITLE
[Translation][Korean] post-06 (edition-2)

### PR DESCRIPTION
**Pull request by JOE1994**
Double checked the translation, and checked rendering using Zola.

Contains some fixes to Korean translations of previous posts (4, 5) to ensure url points to proper sections from previous Korean translations.

### CODE REVIEW

**Comment by phil-opp** [comment]
Thanks a lot!

@Quqqu @SNOOPYOF @dalinaum @KimWang906 Do any of you have time to give this a brief review?

**Comment by dalinaum** [comment]
@phil-opp @JOE1994 I've been busy in the meantime. I'll do a review this weekend. Sorry.

**Comment by phil-opp** [comment]
Thanks a lot! No pressure, only if you have time.

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+[Stack-Segment Fault]: https://wiki.osdev.org/Exceptions#Stack-Segment_Fault
+[General Protection Fault]: https://wiki.osdev.org/Exceptions#General_Protection_Fault
+[Page Fault]: https://wiki.osdev.org/Exceptions#Page_Fault
+
+
+[AMD64 manual]: https://www.amd.com/system/files/TechDocs/24593.pdf
+
+예를 들면 divide-by-zero 예외 뒤에 페이지 폴트가 발생하는 것은 괜찮지만 (페이지 폴트 처리 함수가 호출됨), divide-by-zero 예외 뒤에 general-protection fault 예외가 발생하면 더블 폴트가 발생합니다.
+
+위 테이블을 이용하면 위에서 했던 질문들 중 첫 3개에 대해 대답할 수 있습니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:150]: 질문들 -> 질문

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+3. divide-by-zero 예외 처리 함수가 breakpoint 예외를 일으키는 경우, CPU가 breakpoint 예외 처리 함수의 호출을 시도합니다. breakpoint 예외 처리 함수가 스왑-아웃 되어 있는 경우, _페이지 폴트_ 가 발생하고 _페이지 폴트 처리 함수_ 가 호출됩니다.
+
+사실 임의의 예외에 대한 처리 함수가 IDT에 없다는 것만으로 더블 폴트가 발생하는 것이 아닙니다. 예외가 발생하면 CPU는 그 예외에 대응하는 IDT 엔트리를 참조합니다. 해당 엔트리 값이 0인 경우 (= 예외 처리 함수가 등록되어 있지 않음), _general protection fault_ 예외가 발생합니다. 우리는 해당 예외를 처리할 함수를 등록하지 않았기 때문에, 새로운 general protection fault 예외가 또 발생합니다. general protection fault가 이어서 2번 일어났으니, 위 테이블에 따라 더블 폴트가 발생합니다.
+
+### 커널 스택 오버플로우
+이제 위의 질문들 중 마지막 4번째 질문을 살펴보겠습니다.
+
+> 커널이 스택 오버 플로우를 일으켜 _보호 페이지 (guard page)_ 에 접근하는 경우, 무슨 일이 일어날까요?
+
+보호 페이지는 스택의 맨 아래에 위치하면서 스택 오버플로우를 감지하는 특별한 메모리 페이지입니다. 해당 페이지는 어떤 물리 프레임에도 매핑되지 않으며, CPU가 해당 페이지에 접근하면 물리 메모리에 접근하는 대신 페이지 폴트가 발생합니다. 부트로더가 커널 스택의 보호 페이지를 초기화 하며, 이후 커널 스택 오버플로우가 발생하면 _페이지 폴트_ 가 발생합니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:163]: 초기화 하며 -> 초기화하며

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+
+이러한 스택 교체는 _인터럽트 스택 테이블 (Interrupt Stack Table; IST)_ 을 사용해 진행됩니다. IST는 안전한 것으로 알려진 7개의 다른 스택들의 주소를 저장하는 테이블입니다. IST의 구조를 Rust 코드 형식으로 표현하자면 아래와 같습니다.
+
+```rust
+struct InterruptStackTable {
+    stack_pointers: [Option<StackPointer>; 7],
+}
+```
+
+각 예외 처리 함수는 [IDT 엔트리][IDT entry]의 `stack_pointers` 필드를 통해 IST의 스택들 중 하나를 사용하도록 선택할 수 있습니다. 예를 들어, 우리의 더블 폴트 처리 함수가 IST의 1번째 스택을 사용하도록 설정할 수 있습니다. 그 후에는 더블 폴트가 발생할 때마다 CPU가 스택을 IST의 1번째 스택으로 교체합니다. 스택에 새로운 데이터가 push 되기 전에 스택 교체가 이뤄지기 때문에 트리플 폴트를 피할 수 있습니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:208]: 스택들 중 -> 스택 중

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+
+[IDT entry]: @/edition-2/posts/05-cpu-exceptions/index.md#the-interrupt-descriptor-table
+
+### IST와 TSS
+인터럽트 스택 테이블 (IST)은 오래되어 이젠 구식이 된 _[Task State Segment]_ (TSS)라는 구조체의 일부입니다. 예전에 TSS는 다양한 정보 (예: 프로세서 레지스터들의 상태 값)를 저장하거나 [하드웨어를 이용한 컨텍스트 스위치][hardware context switching]을 지원하는 용도로 사용됐습니다. 하지만 하드웨어를 이용한 컨텍스트 스위치를 64비트 모드에서부터는 지원하지 않게 되었고, 그 이후 TSS의 구조는 완전히 바뀌었습니다.
+
+[Task State Segment]: https://en.wikipedia.org/wiki/Task_state_segment
+[hardware context switching]: https://wiki.osdev.org/Context_Switching#Hardware_Context_Switching
+
+x86_64 아키텍처에서 TSS는 특정 태스크 (task) 관련 정보를 보관하지 않습니다. 대신 TSS는 두 개의 스택 테이블을 보관합니다 (IST가 그 중 하나입니다). 32비트 시스템의 TSS와 64비트 시스템의 TSS의 유일한 공통 필드는 [I/O port permissions bitmap]에 대한 포인터 하나 뿐입니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:218]: 그 중 하나 -> 그중 하나

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+            let stack_start = VirtAddr::from_ptr(unsafe { &STACK });
+            let stack_end = stack_start + STACK_SIZE;
+            stack_end
+        };
+        tss
+    };
+}
+```
+
+Rust의 const evaluator가 위와 같은 TSS의 초기화를 컴파일 중에 진행하지 못해서 `lazy_static`을 사용합니다. IST의 0번째 엔트리가 더블 폴트 스택이 되도록 정합니다 (꼭 0번째일 필요는 없음). 그 다음 더블 폴트 스택의 최상단 주소를 IST의 0번째 엔트리에 저장합니다. 스택의 최상단 주소를 저장하는 이유는 x86 시스템에서 스택은 높은 주소에서 출발해 낮은 주소 영역 쪽으로 성장하기 때문입니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:272]: 그 다음 -> 그다음

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+
+우리가 아직 커널에 메모리 관리 (memory management) 기능을 구현하지 않아서 스택을 할당할 정규적인 방법이 없습니다. 임시 방편으로 `static mut` 배열을 스택 메모리인 것처럼 사용할 것입니다. 값 변경이 가능한 static 변수에 접근하는 경우 컴파일러가 데이터 경쟁 상태 (data race)의 부재를 보장하지 못해 `unsafe` 키워드가 필요합니다. 배열은 꼭 `static`이 아닌 `static mut`로 설정해야 하는데, 그 이유는 부트로더가 `static` 변수를 읽기 전용 메모리 페이지에 배치하기 때문입니다. 이후에 다른 글에서 이 임시적인 스택 메모리 구현을 정석적인 구현으로 수정할 계획이며, 그 후에는 스택 메모리 접근에 더 이상 `unsafe`가 필요하지 않을 것입니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:274]: 임시 방편으로 -> 임시방편으로

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+
+이 더블 폴트 스택에 스택 오버플로우를 감지하기 위한 보호 페이지가 없다는 것에 유의해야 합니다. 더블 폴트 스택에서 스택 오버플로우가 발생하면 스택 아래의 메모리 영역을 일부 덮어쓸 수 있기 때문에, 더블 폴트 처리 함수 안에서 스택 메모리를 과도하게 소모해서는 안됩니다.
+
+#### TSS 불러오기
+새로운 TSS도 만들었으니, 이제 CPU에게 이 TSS를 쓰도록 지시할 방법이 필요합니다. TSS가 역사적 이유로 인해 세그멘테이션 (segmentation) 시스템을 사용하는 탓에, CPU에게 TSS를 쓰도록 지시하는 과정이 꽤 번거롭습니다. TSS를 직접 불러오는 대신, [전역 서술자 테이블 (Global Descriptor Table; GDT)][Global Descriptor Table]을 가리키는 새로운 세그먼트 서술자 (segment descriptor)를 추가해야 합니다. 그 후 [`ltr` 명령어][`ltr` instruction]에 GDT 안에서의 TSS의 인덱스를 주고 호출하여 TSS를 불러올 수 있습니다. (이것이 모듈 이름을 `gdt`로 설정한 이유입니다.)
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:279]: CPU에게 -> CPU에

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+}
+```
+
+이제 GDT를 불러온 상태입니다만 (`_start` 함수가 `init` 함수를 호출했기 때문에), 여전히 커널 스택 오버플로우 발생 시 커널이 무한 재부팅 루프에 갇힙니다.
+
+### 최종 단계
+
+세그먼트 레지스터 및 TSS 레지스터가 기존의 GDT로부터 읽어온 값들을 저장하고 있는 탓에, 우리가 만든 GDT의 세그먼트들이 활성화되지 않은 상황입니다. 또한 더블 폴트 처리 함수가 새로운 스택을 쓰도록 IDT에서 더블 폴트 처리 함수의 엔트리를 알맞게 수정해야 합니다.
+
+정리하자면 우리는 아래의 작업들을 순차적으로 진행해야 합니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:338]: 작업들을 -> 작업을

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+
+수고하셨습니다! 이제부터 더블 폴트가 일어난다면 CPU는 스택을 더블 폴트 스택으로 교체할 것입니다. 드디어 커널 스택 오버플로우가 발생하는 상황을 포함하여 더블 폴트가 일어나는 _어떤 경우라도_ 더블 폴트를 처리할 수 있게 됐습니다.
+
+![QEMU printing `EXCEPTION: DOUBLE FAULT` and a dump of the exception stack frame](qemu-double-fault-on-stack-overflow.png)
+
+앞으로 트리플 폴트를 볼 일은 없을 겁니다! 위에서 구현한 내용을 우리가 미래에 실수로라도 훼손하지 않도록, 위 구현의 작동을 점검하는 테스트를 추가해 보겠습니다.
+
+## 커널 스택 오버플로우 테스트
+
+`gdt` 모듈을 테스트 하고 커널 스택 오버플로우 발생 시 더블 폴트 처리 함수가 호출되는지 확인하는 용도의 통합 테스트를 추가할 것입니다. 테스트 함수에서 더블 폴트를 일으킨 후에 더블 폴트 처리 함수가 호출되었는지 확인하는 테스트를 작성하겠습니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:419]: 테스트 하고 -> 테스트하고

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+}
+
+#[allow(unconditional_recursion)]
+fn stack_overflow() {
+    stack_overflow(); // 재귀 호출할 때마다 반환 주소가 스택에 push 됩니다
+    volatile::Volatile::new(0).read(); // "tail call elimination" 방지하기
+}
+```
+
+`gdt::init` 함수를 호출해 새 GDT를 초기화 합니다. `interrupts::init_idt` 함수 대신 `init_test_idt` 함수를 호출하는데, 그 이유는 패닉하지 않고 `exit_qemu(QemuExitCode::Success)`를 호출하는 새로운 더블 폴트 처리 함수를 등록해 사용할 것이기 때문입니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:485]: 초기화 합니다. -> 초기화합니다

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+
+`stack_overflow` 함수는 `main.rs`에서 작성했던 것과 거의 동일합니다. 유일한 차이점은 함수 마지막에 추가적으로 [`Volatile`] 타입을 이용한 [volatile] 읽기를 통해 [_tail call elimination_]을 방지한다는 것입니다. 주어진 함수의 맨 마지막 구문이 재귀 함수에 대한 호출인 경우, 컴파일러는 tail call elimination 기법을 통해 재귀 함수 호출을 평범한 반복문으로 변환할 수 있습니다. 그렇게 하면 재귀 함수 호출 시 새로운 스택 프레임이 생성되지 않고, 스택 메모리 사용량은 일정하게 유지됩니다.
+
+[volatile]: https://en.wikipedia.org/wiki/Volatile_(computer_programming)
+[`Volatile`]: https://docs.rs/volatile/0.2.6/volatile/struct.Volatile.html
+[_tail call elimination_]: https://en.wikipedia.org/wiki/Tail_call
+
+이 테스트에서 우리는 스택 오버플로우가 발생하기를 원하기 때문에, 함수의 맨 마지막에 컴파일러가 제거할 수 없는 volatile 읽기 작업을 삽입합니다. 따라서 `stack_overflow` 함수는 더 이상 _꼬리 재귀 (tail recursive)_ 함수가 아니게 되고, tail call elimination 기법을 통한 최적화 역시 할 수 없게 됩니다. 또 `allow(unconditional_recursion)` 속성을 함수에 추가해 "함수가 무한히 재귀한다"는 경고 메시지가 출력되지 않게 합니다.
+
+### 테스트 용 IDT
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:495]: 테스트 용 -> 테스트용

File: blog/content/edition-2/posts/06-double-faults/index.ko.md [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+    _stack_frame: InterruptStackFrame,
+    _error_code: u64,
+) -> ! {
+    serial_println!("[ok]");
+    exit_qemu(QemuExitCode::Success);
+    loop {}
+}
+```
+
+더블 폴트 처리 함수가 호출되면 우리는 성공 종료 코드와 함께 QEMU를 종료시키고, 테스트는 성공한 것으로 처리됩니다. 통합 테스트는 완전히 독립적인 실행 파일로 간주되기 때문에, 다시 한 번 테스트 파일의 맨 위에 `#![feature(abi_x86_interrupt)]` 속성을 추가해야 합니다.
```
Comment (dalinaum) [inline blog/content/edition-2/posts/06-double-faults/index.ko.md:545]: 간주되기 -> 간주하기

**Comment by dalinaum** [comment]
@JOE1994 @phil-opp I added subtle comments. Good job.

**Comment by JOE1994** [comment]
@dalinaum
I've pushed a commit to apply your feedback and to add you as a contributor to the translation page.
Thank for your time!

@phil-opp
I think this PR is ready for your review :+1:

**Review by phil-opp — APPROVED** [review]

**Pull request merged**